### ISSUE TITLE
End of 2023 random thoughts

### ISSUE THREAD

**Issue by alecgunny** [alecgunny/ligo-handoff-resources#1]
- When installing projects in singularity definition files, to avoid issues with updates to `virtualenv` breaking `poetry`, you can use the [`poetry export`](https://python-poetry.org/docs/cli/#export) command to export the dependencies to a `requirements.txt`, then just do a `pip install -r` on that file (in fact, if you were doing this with Docker, you could do this as a separate build stage and then just copy the `requirements.txt` to the production stage, and never have to fuss with poetry in the final image at all). I don't know if poetry exports local dependencies as editable in the requirements file, but if not it should be trivial to use `sed` or something to look for local deps and prepend a `-e` flag to them. The other thing to keep in mind is that `export` doesn't include all groups by default, so you'll probably want to include `--with dev` to get dev dependencies in there
- Here's a good [hyperparameter tuning playbook](https://github.com/google-research/tuning_playbook) from some Google researchers that was shared with me recently. Really cool to see how many of their tips aframe adopted by our own experimentation.
- The existing `aframe` code [always adds each step's loss](https://github.com/ML4GW/aframe/blob/3f4514969a0f8d307ea68bc648839955497ea2c6/libs/trainer/aframe/trainer/trainer.py#L35) to the running `train_loss`, even if it's `nan` and will be rejected by the `GradientScaler`. Since we only report epoch-level losses, it's possible that you could be seeing `nan`s here even if training is proceeding reasonably well. More often than not, it won't be, but it would be good practice when using mixed-precision training to only add things to the `train_loss` if they aren't rejected by the scaler.

**Comment by EthanMarx** [comment]
Thanks for the thoughts!
1. That poetry tip is nice, will take a look at implementing that eventually.
2. I actually bumped into that google hp document over the holidays, seems extremely useful will have to give it an end to end
3. Im guessing when we move to aframev2, lightning will take care of this?

Hope you're enjoying bluerose

**Comment by alecgunny** [comment]
@EthanMarx on (3) - yeah exactly. Reason number 1,002,891 to not implement your own training step 😑
